bemserver_core/database.py

```diff
@@ -174,3 +188,61 @@ def create_all(self, **kwargs):
 
 
 db = DBConnection()
+
+
+# Inspired by https://dba.stackexchange.com/a/61304 for the use of `to_json` function.
+def init_db_functions():
+    """Create functions for triggers...
+
+    This function is meant to be used for tests or dev setups after create_all.
+    Production setups should rely on migration scripts.
+    """
+
+    # SQL function to raise an exception when trying to update a "read-only" column.
+    db.session.execute(
+        sqla.DDL(
+            dedent(
+                """\
+                CREATE FUNCTION column_update_not_allowed()
+                    RETURNS TRIGGER AS
+                $func$
+                    DECLARE
+                        col_name text := TG_ARGV[0]::text;
+                    BEGIN
+                        RAISE EXCEPTION USING
+                            MESSAGE = col_name || ' cannot be modified',
+                            ERRCODE = 'integrity_constraint_violation';
+                    END;
+                $func$
+                LANGUAGE plpgsql STRICT;\
+                """
+            )
+        )
+    )
+
+    db.session.commit()
+
+
+def generate_ddl_trigger_readonly(table_name, col_name):
+    """Generate the SQL statement that creates an "update read-only trigger"
+    on a specific column for a table.
+
+    :param str table_name: The name of the table concerned by the trigger.
+    :param str col_name: The name of the column to protect from update.
+    :returns sqlalchemy.DDL: An instance of DDL statement that creates the trigger.
+    """
+    return sqla.DDL(
+        dedent(
+            f"""\
+            CREATE TRIGGER
+                {table_name}_trigger_update_readonly_{col_name}
+            BEFORE UPDATE
+                OF {col_name} ON {table_name}
+            FOR EACH ROW
+                WHEN (
+                    NEW.{col_name} IS DISTINCT FROM OLD.{col_name}
+                )
+                EXECUTE FUNCTION column_update_not_allowed({col_name});\
```

Comment:
Wondering (out loud) if at this point the `column_update_not_allowed` could be inlined here. Looks like half of it is boilerplate.

Reply:
Apparently, this is the way triggers work. Function must be declared separately.